File: Competitive Programming 3/Section 1.4/Anagram/630 - Anagrams (II)/C++/main.cpp
#include<bits/stdc++.h>

using namespace std;

int main(){
    ios_base::sync_with_stdio(0);cin.tie(0);
    int t,n;
    string line;
    cin>> t;
    cin.ignore();
    for(int cs=0;cs<t;cs++){
        map<string,vector<string> > anagrams;
        if(cs!=0)
            cout << "\n";
        cin.ignore();
        cin>>n;
        cin.ignore();
        //build anagram ds
        while(n--){
            getline(cin,line);
            string sortedLine = line;
            sort(sortedLine.begin(),sortedLine.end());
            anagrams[sortedLine].push_back(line);
        }
        //anagram queries
        while(getline(cin,line)&&line!="END"){
            string original = line;
            cout << "Anagrams for: " << original << "\n";
            sort(line.begin(),line.end());
            auto it = anagrams.find(line);
            if(it==anagrams.end())
                cout << "No anagrams for: " << original << "\n";
            else{
                auto vectorRef = it->second;
                for(size_t i = 0 ; i < vectorRef.size(); i++)
                    cout << setw(3) << i+1 << ") " << vectorRef[i] << "\n";
            }
        }
    }
    return 0;
}

File: Competitive Programming 3/Section 1.4/Anagram/630 - Anagrams (II)/Python/main.py
def main():
    t = int(input())
    for cs in range(t):
        if cs!=0:
            print()
        anagrams={}
        input()
        #build anagram ds
        for _ in range(int(input())):
            line = input()
            sortedLine = ''.join(sorted(line))
            if sortedLine not in anagrams:
                anagrams[sortedLine]=[]
            anagrams[sortedLine].append(line)
        #anagrams queries
        while True:
            line = input()
            if line == 'END':
                break
            print('Anagrams for:',line)
            sortedLine = ''.join(sorted(line))
            if sortedLine in anagrams:
                listRef = anagrams[sortedLine]
                for i in range(len(listRef)):
                    print('{0:>3}'.format(i+1)+')',listRef[i])
            else:
                print('No anagrams for:',line)
main()